### PR TITLE
Add event listener support to the Desktop Agent API

### LINKED ISSUE #1136: Add event listeners (for non-context events) to the Desktop Agent API (e.g. for changes to the current User channel)
### Use Case:
Typical Desktop Agent implementations for containers allow the end user to set the current User Channel for an app through an appropriate UX. However, applications may also implement their own channel management UX, by gathering details of the channels with `fdc3.getUserChannels` and effecting changes by calling the `fdc3.joinUserChannel` and `fdc3.leaveCurrentChannel`. There is currently no support for listening to changes to the current channel setting - which may be made outside of the application (e.g. via a channel selection menu on the window titlebar or other chrome), making it difficult to update any built-in UI for changes made. To help these use cases to work together, and to enable other use cases where the application's behaviour changes once it is linked to a channel, an event listener for channel changes is needed.

Additionally, (and as discussed at #1106) it is anticipated that FDC3 for the web will increase the need for the ability to listen for changes, as channel selector might be implemented in a number of forms (e.g. floating channel selector rendered within the window, in an outer window (where the application is embedded in an ifrmae within the window), etc..

As discussed at #1160 and #1196, the current proposal is to add a generic `addEventListener` function, for future-proofing against future events (other than context and intent listeners which have their own API signatures) being added to the FDC3 API. A number of FDC3 participants noted that connection events may be useful for FDC3 in a Web browser (loss of Connection, DA was closed etc.).

Proposed function:

```ts
/** Register a handler for events from the Desktop Agent. Whenever the handler function 
 * is called it will be passed an event object with details related to the event.
 * @parm {FDC3EventType  | null} type If non-null, only events of the specified type will be received by the handler. 
 * @param {EventHandler} handler A function that events received will be passed to. 
 */
addEventListener(type: FDC3EventType  | null, handler: EventHandler): Promise<Listener>;

/** Type representing a handler function for events from the Desktop Agent. 
 * @param {FDC3Event} event The handler function will be passed an `FDC3Event` Object 
 * providing details of the event (such as a change of channel membership for the app) as the only 
 * parameter.
 */
type EventHandler = (event: FDC3Event) => void;

/**
 * Enumeration defining the types of (non-context and non-intent) events that may be received
   via the FDC3 API's `addEventListener` function. 
 */
enum FDC3EventType = {
  USER_CHANNEL_CHANGED = "USER_CHANNEL_CHANGED"
}

/**
 * Type representing the format of event objects that may be received
   via the FDC3 API's `addEventListener` function. 
 */
type FDC3Event = {
  readonly type: FDC3EventType;
  readonly details: any;
}

/**
 * Type representing the format of event USER_CHANNEL_CHANGED objects
 */
type FDC3ChannelChangedEvent extends FDC3Event {
  readonly type: FDC3EventType.USER_CHANNEL_CHANGED;
  readonly details: {
    currentChannelId: string | null
  };
}
```

### PR DESCRIPTION
**Pull request by kemerava**
Closes #1136 
Adding support for event listener for non-context and non-intent events 
@michael-bowen-sc

### CODE REVIEW

File: docs/api/ref/Events.md
Comment (kriswest) [inline docs/api/ref/Events.md:59]:
```suggestion
Type representing the format of USER_CHANNEL_CHANGED events. The identity of the channel joined is provided as `details.currentChannelId`, which will be `null` if the app is no longer joined to any channel.
```

**Review by kriswest — COMMENTED** [review]
@kemerava Many thanks for the contribution here! I've made a few suggestions to tighten it up.

Please note the comment from EasyCLA, we'll need you to click on the authorization link and to make sure your firm has a CCLA in place (contact fdc3-maintainers@finos.org and help@finos.org if you have any issues there).  I also note one of the commits has an issue with a missing github id that will also need to be fixed or the PR recreated without it.

**Comment by kemerava** [comment]
Thank you for the review, @kriswest!
I am working on trying to figure out why we are getting the error from the CCLA side, as we have indeed signed the CCLA already.
Regarding that commit with missing github id, I was able to resolve that, and the error message went away. 
Thank you for the suggestions, I have applied them, however, right after that easyCLA posted the same issue as before with the missing id. On the commit it has just you and I as authors:
Screenshot 2024-05-17 at 1 57 06 PM
Do you know why this is causing an issue, as both of our handles seem to be recognized?

**Comment by kriswest** [comment]
> I am working on trying to figure out why we are getting the error from the CCLA side, as we have indeed signed the CCLA already.

If the CCLA is in place, you still need to click on the [Please click here to be authorized](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/13399143/121826936/1207/#/?version=2) link - that will confirm your status on the CCLA and future updates won't require it. In case you've already done that, I'll see if I can get easyCLA to recheck via a comment - failing that we may have to ask their support for help (or further assistance with EasyCLA, [please submit a support request ticket](https://jira.linuxfoundation.org/servicedesk/customer/portal/4).)

Glad you got the other error resolved!

**Comment by kemerava** [comment]
Thank you, @kriswest!
I created the support issue for the github account easyCLA error: https://jira.linuxfoundation.org/plugins/servlet/desk/portal/4/SUPPORT-26917

Thanks for the info on the CCLA part, I will bring it back to my team to let them know

**Comment by kriswest** [comment]
I opened one as well @kemerava: https://jira.linuxfoundation.org/plugins/servlet/desk/portal/4/SUPPORT-26910

They may know what's up... but also made a comment that suggests you haven't yet hit the 'click here to be authorized' link in the comment the EasyCLA bot made above and selected proceed as a corporate contributor and been approved - if you have, add a comment here so they see it.

**Comment by kriswest** [comment]
Also if you visit: https://patch-diff.githubusercontent.com/raw/finos/FDC3/pull/1207.patch you'll see  the emial it is picking up for you (which may not be the right one to get you approved on a corporate CLA)

**Comment by kriswest** [comment]
@kemerava they rolled back easyCLA and it is running checks again properly - it cleared one of the errors so I know its updated. Now it just needs to know which CCLA you are on - so go hit that [Please click here to be authorized](https://api.easycla.lfx.linuxfoundation.org/v2/repository-provider/github/sign/13399143/121826936/1207/#/?version=2) link and complete the steps. If you still get nowhere then maybe add your work email to your github account if you haven't done so already - the email domain is often used as the approval criteria by CLA managers. Alternatively, your firm's CLA manager can manually add you to the CCLA.

**Comment by kemerava** [comment]
Hi @kriswest,

Just finished figuring out the CLA from our side. Looks like there are no more errors from the easyCLA side. 
Please let me know what else is required from my side for this PR.
Thank you for all your help!

**Comment by kriswest** [comment]
@kemerava looks good to me - we got consent at today's meeting to go ahead with this once it passes a couple of reviews, which will hopefully happen over the next couple of weeks.

Again, many thanks for the contribution!

File: docs/api/ref/DesktopAgent.md
Comment (hughtroeger) [inline docs/api/ref/DesktopAgent.md:100]:
```suggestion
```ts
addEventListener(type: FDC3EventType  | null, handler: EventHandler): Promise<Listener>;
```
```
Currently has quadruple backticks

File: docs/api/ref/Events.md
Comment (hughtroeger) [inline docs/api/ref/Events.md:40]:
```suggestion
```
```

File: docs/api/ref/Events.md
Comment (hughtroeger) [inline docs/api/ref/Events.md:57]:
```suggestion
```ts
interface FDC3ChannelChangedEvent extends FDC3Event {
  readonly type: FDC3EventType.USER_CHANNEL_CHANGED;
  readonly details: {
    currentChannelId: string | null
  };
}
```
```

Comment (hughtroeger) [reply docs/api/ref/Events.md:57]: quadruple backticks

File: src/api/DesktopAgent.ts
Comment (hughtroeger) [inline src/api/DesktopAgent.ts:392]:
```suggestion
   * ```js
   * // any event type
   * const listener = await fdc3.addEventListener(null, event => { ... });
   * 
   * // listener for a specific event type that logs its details
   * const userChannelChangedListener = await fdc3.addEventListener(FDC3EventType.USER_CHANNEL_CHANGED, event => { 
   * console.log(`Received event ${event.type}\n\tDetails: ${event.details}`);
   * //do something else with the event
   * });
   * ```
```

**Review by hughtroeger — COMMENTED** [review]
All comments are nitpicks about backticks. Content looks good, matches the proposed API in the issue.

File: docs/api/ref/DesktopAgent.md
Comment (kriswest) [reply docs/api/ref/DesktopAgent.md:100]: @hughtroeger FYI this suggestion was dropping the last line due a clash between the suggestion syntax and the triple backticks in what was being corrected. To avoid that use quadruple backticks in the suggestion!

**Comment by kriswest** [comment]
LGTM and matches the proposal. Applied your suggestions @hughtroeger after figuring out that you need to use quadruple backticks in github suggestions involving triple backticks!

I was involved in the proposal, so adding other maintainers to review and will leave merging a week or two to facilitate.

**Comment by kemerava** [comment]
Hi @kriswest, any updates on this? Anything needed from my side? Thank you!

**Comment by kriswest** [comment]
Hi @kemerava, nothing more is needed apart from a review from another maintainer (which it looks like I forgot to request - done now)!

**Review by mistryvinay — APPROVED** [review]
LGTM.

**Comment by kemerava** [comment]
Hi @kriswest,
Just checking in here, are there any more approvals required for this? 
Thank you so much for helping with this PR!
Elizabeth

**Review by bingenito — APPROVED** [review]

**Comment by michael-bowen-sc** [comment]
> Hi @kriswest, Just checking in here, are there any more approvals required for this? Thank you so much for helping with this PR! Elizabeth

Looks like this is all green.  What is holding this PR up from merge?

**Comment by kriswest** [comment]
@michael-bowen-sc  now that Brian and Vinay have had a look it's all good to go. Apologies for the delay, OSFF prep was.distractong me!

**Comment by kriswest** [comment]
This PR when in without a changelog entry, which we will need to add before the next version of FDC3 is adopted

**Pull request merged**
